File: pfa4-main/pfa4-main/amn-frontend/src/app/services/PatientService.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface PatientBasicInfoDTO {
  id: number;
  fullName: string;
  email: string;
  cin: string;
  address: string;
  phone: string;
  bloodType: string;
}

export interface PatientProfileDTO {
  id: number;
  fullName: string;
  cin: string;
  bloodType: string;
  birthDate: string;
  emergencyContact: string;
  allergies: string;
  chronicDiseases: string;
  hasHeartProblem: boolean;
  hasSurgery: boolean;
  email: string;
  phone: string;
  address: string;
  medicalFolderId: number;

  // Collections
  medicalRecords: MedicalRecordDTO[];
  scans: ScanDTO[];
  analyses: AnalysisDTO[];
  surgeries: SurgeryDTO[];
  vaccinations: VaccinationDTO[];
  prescriptions: PrescriptionDTO[];
}

export interface MedicalRecordDTO {
  id: number;
  reason: string;
  diagnosis: string;
  notes: string;
  creationDate: string;
  doctorName: string;
  prescriptions?: PrescriptionDTO[];
}

export interface PrescriptionDTO {
  id: number;
  prescribedDate: string;
  status: string;
  prescribingDoctor: string;
  medications: MedicationDTO[];
  medicalRecordId?: number;
}

export interface MedicationDTO {
  id: number;
  name: string;
  dosage: string;
  period: string;
  permanent: boolean;
}

export interface ScanDTO {
  id: number;
  title: string;
  description: string;
  uploadDate: string;
  url: string;
}

export interface AnalysisDTO {
  id: number;
  title: string;
  description: string;
  uploadDate: string;
  url: string;
}

export interface SurgeryDTO {
  id: number;
  title: string;
  description: string;
  uploadDate: string;
  url: string;
}

export interface VaccinationDTO {
  id: number;
  vaccineName: string;
  doseNumber: number;
  manufacturer: string;
  vaccinationDate: string;
}

@Injectable({
  providedIn: 'root',
})
export class PatientService {
  private apiUrl = 'http://localhost:8080/api/patient';

  constructor(private http: HttpClient) {}

  /**
   * Get JWT token from localStorage
   */
  private getJwt(): string | null {
    return localStorage.getItem('jwt');
  }

  /**
   * Get authorization headers
   */
  private getHeaders(): HttpHeaders {
    const token = this.getJwt();
    let headers = new HttpHeaders();
    if (token) {
      headers = headers.set('Authorization', `Bearer ${token}`);
    }
    return headers;
  }

  /**
   * ✅ Get complete patient profile by email
   */
  getPatientProfileByEmail(email: string): Observable<PatientProfileDTO> {
    const params = new HttpParams().set('email', email);
    return this.http.get<PatientProfileDTO>(`${this.apiUrl}/profile-by-email`, {
      params,
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Get complete patient profile by CIN
   */
  getPatientProfileByCin(cin: string): Observable<PatientProfileDTO> {
    const params = new HttpParams().set('cin', cin);
    return this.http.get<PatientProfileDTO>(`${this.apiUrl}/profile-by-cin`, {
      params,
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Get patient profile by folder ID
   */
  getPatientProfileByFolderId(folderId: number): Observable<PatientProfileDTO> {
    return this.http.get<PatientProfileDTO>(`${this.apiUrl}/profile-by-folder/${folderId}`, {
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Get current patient basic data
   */
  getCurrentPatient(email: string): Observable<{cin: string, email: string}> {
    const params = new HttpParams().set('email', email);
    return this.http.get<{cin: string, email: string}>(`${this.apiUrl}/current`, {
      params,
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Fetch basic patient info by CIN
   */
  getPatientBasicInfo(cin: string): Observable<PatientBasicInfoDTO> {
    const params = new HttpParams().set('cin', cin);
    return this.http.get<PatientBasicInfoDTO>(`${this.apiUrl}/basic-info`, {
      params,
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Update patient basic info
   */
  updatePatientBasicInfo(cin: string, basicInfo: PatientBasicInfoDTO): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/basic-info/${cin}`, basicInfo, {
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Get medical records for a patient by folder ID
   */
  getMedicalRecords(folderId: number): Observable<MedicalRecordDTO[]> {
    return this.http.get<MedicalRecordDTO[]>(`${this.apiUrl}/records/${folderId}`, {
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Get prescriptions for a patient by folder ID
   */
  getPrescriptions(folderId: number): Observable<PrescriptionDTO[]> {
    return this.http.get<PrescriptionDTO[]>(`${this.apiUrl}/prescriptions/${folderId}`, {
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Get scans for a patient by folder ID
   */
  getScans(folderId: number): Observable<ScanDTO[]> {
    return this.http.get<ScanDTO[]>(`${this.apiUrl}/scans/${folderId}`, {
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Get analyses for a patient by folder ID
   */
  getAnalyses(folderId: number): Observable<AnalysisDTO[]> {
    return this.http.get<AnalysisDTO[]>(`${this.apiUrl}/analyses/${folderId}`, {
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Get surgeries for a patient by folder ID
   */
  getSurgeries(folderId: number): Observable<SurgeryDTO[]> {
    return this.http.get<SurgeryDTO[]>(`${this.apiUrl}/surgeries/${folderId}`, {
      headers: this.getHeaders()
    });
  }

  /**
   * ✅ Get vaccinations for a patient by folder ID
   */
  getVaccinations(folderId: number): Observable<VaccinationDTO[]> {
    return this.http.get<VaccinationDTO[]>(`${this.apiUrl}/vaccinations/${folderId}`, {
      headers: this.getHeaders()
    });
  }
}

File: pfa4-main/pfa4-main/src/main/java/com/amn/service/PatientService.java
package com.amn.service;

import com.amn.dto.*;
import com.amn.entity.*;
import com.amn.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class PatientService {

    private final PatientRepository patientRepository;
    private final MedicalFolderRepository medicalFolderRepository;
    private final AnalysisRepository analysisRepository;
    private final ScanRepository scanRepository;
    private final SurgeryRepository surgeryRepository;
    private final PrescriptionRepository prescriptionRepository;
    private final MedicalRecordRepository medicalRecordRepository;
    private final VaccinationRepository vaccinationRepository;
    private final UserRepository userRepository;

    /**
     * ✅ Get Patient Profile by CIN
     */
    @Transactional(readOnly = true)
    public PatientProfileDTO getPatientProfileByCin(String cin) {
        // Get just the basic patient data first
        Patient patient = patientRepository.findByCin(cin)
                .orElseThrow(() -> new RuntimeException("Patient not found for CIN: " + cin));

        return buildPatientProfileFixed(patient);
    }

    /**
     * ✅ Get Patient Profile by Folder ID
     */
    @Transactional(readOnly = true)
    public PatientProfileDTO getPatientProfileByFolderId(Long folderId) {
        // Get just the folder first
        MedicalFolder folder = medicalFolderRepository.findById(folderId)
                .orElseThrow(() -> new RuntimeException("Medical Folder not found for ID: " + folderId));

        // Then get the patient separately
        Patient patient = patientRepository.findById(folder.getPatient().getId())
                .orElseThrow(() -> new RuntimeException("No patient associated with folder ID: " + folderId));

        return buildPatientProfileFixed(patient);
    }

    /**
     * ✅ Get Patient by Email
     */
    @Transactional(readOnly = true)
    public Patient getCurrentPatient(String email) {
        return patientRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("Patient not found with email: " + email));
    }

    /**
     * ✅ Fetch Patient Profile by Email
     */
    @Transactional(readOnly = true)
    public PatientProfileDTO getPatientProfileByEmail(String email) {
        Patient patient = getCurrentPatient(email);
        return buildPatientProfileFixed(patient);
    }

    /**
     * ✅ FIXED: Build Patient Profile DTO to avoid MultipleBagFetchException
     */
    private PatientProfileDTO buildPatientProfileFixed(Patient patient) {
        // Start by building the basic profile without any collections
        PatientProfileDTO profile = PatientProfileDTO.builder()
                .id(patient.getId())
                .fullName(patient.getFullName())
                .cin(patient.getCin())
                .bloodType(patient.getBloodType())
                .birthDate(String.valueOf(patient.getBirthDate()))
                .emergencyContact(patient.getEmergencyContact())
                .allergies(patient.getAllergies())
                .chronicDiseases(patient.getChronicDiseases())
                .hasHeartProblem(patient.isHasHeartProblem())
                .hasSurgery(patient.isHasSurgery())
                .email(patient.getEmail())
                .phone(patient.getPhone())
                .address(patient.getAddress())
                .build();

        // Get folder ID in a separate query
        Optional<MedicalFolder> folderOpt = medicalFolderRepository.findByPatientId(patient.getId());
        if (!folderOpt.isPresent()) {
            return profile; // Return just the basic profile if no folder exists
        }

        Long folderId = folderOpt.get().getId();
        profile.setMedicalFolderId(folderId);

        // Load each collection with separate transactions to avoid MultipleBagFetchException
        // This works because each method issues its own separate SQL query
        profile.setMedicalRecords(getMedicalRecords(folderId));
        profile.setPrescriptions(getPrescriptions(folderId));
        profile.setScans(getScans(folderId));
        profile.setAnalyses(getAnalyses(folderId));
        profile.setSurgeries(getSurgeries(folderId));
        profile.setVaccinations(getVaccinations(folderId));

        return profile;
    }

    /**
     * ✅ Fetch Medical Records by Folder ID
     */
    @Transactional(readOnly = true)
    public List<MedicalRecordDTO> getMedicalRecords(Long folderId) {
        return medicalRecordRepository.findByMedicalFolderId(folderId).stream()
                .map(MedicalRecordDTO::fromEntity)
                .collect(Collectors.toList());
    }

    /**
     * ✅ Fetch Prescriptions by Folder ID
     */
    @Transactional(readOnly = true)
    public List<PrescriptionDTO> getPrescriptions(Long folderId) {
        return prescriptionRepository.findByMedicalFolderId(folderId).stream()
                .map(PrescriptionDTO::fromEntity)
                .collect(Collectors.toList());
    }

    /**
     * ✅ Fetch Scans by Folder ID
     */
    @Transactional(readOnly = true)
    public List<ScanDTO> getScans(Long folderId) {
        return scanRepository.findByMedicalFolderId(folderId).stream()
                .map(ScanDTO::fromEntity)
                .collect(Collectors.toList());
    }

    /**
     * ✅ Fetch Analyses by Folder ID
     */
    @Transactional(readOnly = true)
    public List<AnalysisDTO> getAnalyses(Long folderId) {
        return analysisRepository.findByMedicalFolderId(folderId).stream()
                .map(AnalysisDTO::fromEntity)
                .collect(Collectors.toList());
    }

    /**
     * ✅ Fetch Surgeries by Folder ID
     */
    @Transactional(readOnly = true)
    public List<SurgeryDTO> getSurgeries(Long folderId) {
        return surgeryRepository.findByMedicalFolderId(folderId).stream()
                .map(SurgeryDTO::fromEntity)
                .collect(Collectors.toList());
    }

    /**
     * ✅ Fetch Vaccinations by Folder ID
     */
    @Transactional(readOnly = true)
    public List<Vaccination> getVaccinations(Long folderId) {
        return vaccinationRepository.findByMedicalFolderId(folderId);
    }

    /**
     * ✅ Get Patient Basic Info
     */
    @Transactional(readOnly = true)
    public PatientBasicInfoDTO getPatientBasicInfo(String cin) {
        return patientRepository.findPatientBasicInfoByCin(cin);
    }

    /**
     * ✅ FIXED: Update Patient Basic Info (Address, Email, Phone)
     */
    @Transactional
    public void updatePatientBasicInfo(String cin, PatientBasicInfoDTO basicInfo) {
        // Use JPQL to get only the patient entity without collections
        Patient patient = patientRepository.findByCin(cin)
                .orElseThrow(() -> new RuntimeException("Patient not found for CIN: " + cin));

        // Get User separately without collections
        User user = userRepository.findById(patient.getId())
                .orElseThrow(() -> new RuntimeException("User not found for Patient ID: " + patient.getId()));

        // Update Address
        if (basicInfo.getAddress() != null) {
            user.setAddress(basicInfo.getAddress());
        }

        // Update Email
        if (basicInfo.getEmail() != null) {
            // Check if email exists but make sure we're not checking against the same user
            boolean emailExists = userRepository.existsByEmail(basicInfo.getEmail());
            if (emailExists && !basicInfo.getEmail().equals(user.getEmail())) {
                throw new RuntimeException("Email is already in use.");
            }
            user.setEmail(basicInfo.getEmail());
        }

        // Update Phone
        if (basicInfo.getPhone() != null) {
            user.setPhone(basicInfo.getPhone());
        }

        // Save the User entity
        userRepository.save(user);
    }
}